linkerd/protocol/h2/src/main/scala/io/buoyant/linkerd/protocol/h2/H2RequestActiveTracer.scala

```diff
@@ -0,0 +1,147 @@
+ * @param namers namers used to evaluate a service name
+ * @param dtab base dtab used by the router
+ * @param routerLabel name of router
+ */
+class H2RequestActiveTracer(
+  endpoint: EndpointAddr,
+  namers: DstBindingFactory.Namer,
+  dtab: BaseDtab,
+  routerLabel: String
+) extends SimpleFilter[Request, Response] {
+
+  private[this] val MaxForwards = "max-forwards"
+  private[this] val AddRouterContextHeader = "l5d-add-context"
+
+  private[this] def getRequestTraceResponse(resp: Response, stopwatch: Stopwatch.Elapsed):Future[Response] = {
+    val routerCtxF = RouterContextBuilder(
+      routerLabel,
+      stopwatch,
+      DstPathCtx.current,
+      DstBoundCtx.current,
+      endpoint,
+      namers,
+      dtab)
+
+    routerCtxF.map { ctx =>
+      // format current router context
+      val content = s"""|
+                        |${ctx.formatRouterContext.trim}
+                        |""".stripMargin
+      if (resp.stream.isEmpty) {
+        // add router context data frame to empty stream
+        Response(resp.headers, Stream.const(Frame.Data(content, eos = true)))
+      } else {
+        val newStream = resp.stream.flatMap {
+          // read every frame of a stream
+          case trailerFrame: Trailers =>
+            // on trailers frame, prepend a data frame containing router context to trailer
+            val routerCtxFrame = Frame.Data(content, eos = false)
+            Seq(routerCtxFrame, trailerFrame)
+          case dataFrame: Data if dataFrame.isEnd =>
+            // on data frame with EOS flag to true, prepend new data frame with router context to
+            // EOS data frame
+            val newDataFrame = Frame.Data.copy(dataFrame, eos = false)
+            val routerCtxFrame = Data(content, eos = true)
+            Seq(newDataFrame, routerCtxFrame, Trailers())
+          case frame =>
+            // do nothing if we find a frame that doesn't indicate the end of a stream
+            Seq(frame)
+        }
+        Response(resp.headers, newStream)
+      }
+    }
+
+  }
+
+  override def apply(
+    request: Request,
+    service: Service[Request, Response]
+  ): Future[Response] = {
+    val maxForwards = Try(request.headers.get(MaxForwards).map(_.toInt))
+    val isAddRouterCtx = request.headers.get(AddRouterContextHeader) match {
+      case Some(v) => StringUtil.toBoolean(v)
+      case None => false
+    }
+
+    (request.method, maxForwards, isAddRouterCtx) match {
+      case (Method.Trace, Throw(_: NumberFormatException), _) =>
+        // MaxForwards is unparsable send a bad request response
+        Future.value(Response(Status.BadRequest, Stream.empty))
+      case (Method.Trace, Return(Some(0)), true) =>
+        // MaxForwards header is 0 and the l5d-add-context header is present
+        // returns a response with router context
+        val stopwatch = Stopwatch.start()
+        getRequestTraceResponse(Response(request.headers, Stream.empty), stopwatch)
```

Comment:
I think this is invalid.  I'm guessing that the response needs to have certain response headers (like status code) and cannot have certain request headers (like authority, path, method, etc).

Reply:
Ahh! good point.